ntbs-service/Pages/Notifications/Edit/Children/SocialContextVenue.cshtml.cs

```diff
@@ -0,0 +1,185 @@
+            }
+
+            var venue = Notification.SocialContextVenues
+                    .SingleOrDefault(s => s.SocialContextVenueId == RowId.Value);
+            if (venue == null)
+            {
+                return NotFound();
+            }
+
+            await _socialContextVenueRepository.DeleteAsync(venue);
+
+            return RedirectToPage("/Notifications/Edit/SocialContextVenues", new { NotificationId });
+        }
+
+        private async Task SetDropdownsAsync()
+        {
+            var venueTypes = await _referenceDataRepository.GetAllVenueTypesAsync();
+            VenueTypes = new SelectList(
+                items: venueTypes,
+                dataValueField: nameof(VenueType.VenueTypeId),
+                dataTextField: nameof(VenueType.Name),
+                selectedValue: null,
+                dataGroupField: nameof(VenueType.Category));
+        }
+
+        protected override IActionResult RedirectAfterSaveForNotified()
+        {
+            return RedirectToPage("/Notifications/Edit/SocialContextVenues", new { NotificationId });
+        }
+
+        protected override IActionResult RedirectAfterSaveForDraft(bool isBeingSubmitted)
+        {
+            return RedirectToPage("/Notifications/Edit/SocialContextVenues", new { NotificationId, isBeingSubmitted });
+        }
+
+        protected override async Task<Notification> GetNotificationAsync(int notificationId)
+        {
+            return await NotificationRepository.GetNotificationWithSocialContextVenuesAsync(notificationId);
+        }
+
+
+        private void SetDates()
+        {
+            // The required date will be marked as missing on the model, since we are setting it manually, rather than binding it
+            ModelState.Remove("Venue.DateFrom");
+            ValidationService.TrySetFormattedDate(Venue, "Venue", nameof(Venue.DateFrom), FormattedDateFrom);
+            ModelState.Remove("Venue.DateTo");
+            ValidationService.TrySetFormattedDate(Venue, "Venue", nameof(Venue.DateTo), FormattedDateTo);
+        }
+
+        private async Task FindAndSetPostcodeAsync()
+        {
+            ModelState.ClearValidationState("Venue.Postcode");
+            await FindAndSetPostcodeAsync(_postcodeService, Venue);
+        }
+
+        public async Task<ContentResult> OnGetValidatePostcode(string postcode, bool shouldValidateFull)
+        {
+            return await OnGetValidatePostcode<SocialContextVenue>(_postcodeService, postcode, shouldValidateFull);
+        }
+
+        public ContentResult OnGetValidateVenueProperty(string key, string value, bool shouldValidateFull)
+        {
+            return ValidationService.ValidateModelProperty<SocialContextVenue>(key, value, shouldValidateFull);
+        }
+
+        public ContentResult OnGetValidateVenueDate(string key, string day, string month, string year)
+        {
+            return ValidationService.ValidateDate<SocialContextVenue>(key, day, month, year);
+        }
+

```

Comment:
This is me starting to implement the dynamic validation of checking that the two date fields (from and to) are in the correct "order" (i.e. To has to be after From). I initially tried to reuse the existing "ValidateMultiple" component, which I thought would be easy to extend -and it would be for a normal input fields, but as date fields are split into 3 fields it's quite a bit more complex.

We already do comparison of dates in the DateComparison component used on ClinicalDetails (which is fully client side) - however there we are showing only warnings, and they aren't displayed in the same way as usual error messages....

Before proceeding I thought it would be best to check what we should do, as either route will take quite a bit of time to get to work (the validation obviously works server-side, so we might be able to already push this and then do the dynamic bit as a tech debt ticket as this whole ticket is taking quite a while...).

I feel like if we want to reuse the datecomparison component, it would be important to style it such that it looks like a normal error message - and maybe even completely refactor it so that for ClinicalDetails the warnings look like normal error messages, just in a different colour.
Alternatively, I could leave datecomparison as is and try to extend ValidateMultiple. I definitely don't think we should write a completely new component, as the existing ones solve a lot of the issues we need to solve here. What do you think?

Reply:
The warnings are styled according to GDS, and I don't think different coloured validation style messages would be a great UX.
We could hypothetically have an inner component which hits the db and emits a message to an outer 'display' component?

Reply:
Ok, I agree we shouldn't restyle it if we are following a GDS component. However, the date comparison warnings at the moment are not very clearly placed - they just appear in one place at the top, and it would probably be better for them to be next to the actual field that is being validated. Then it should hopefully be easily extendible to the proper error case.

Basically, to me choosing between extending validatemultiple vs datecomparison boils down to this:

**ValidateMultiple**

Pros: 

- Makes a call to razor page like all other dynamic validation components

Cons:

- Will need to re-implement dealing with date fields which is a faff

**DateComparison**

Pros: 

- Can already handle date fields and compare them
- Could take this as an opportunity to make the UI for clinical details warnings nicer too, by placing warnings next to affected fields

Cons:

- Would do validation of these fields in javascript only, which differs from all other cases

Reply:
Sorry only getting around to chiming in now! I think there may be a case for a third option in not doing this validation dynamically. If that gets shut down by Nancy, then I recon we should attempt to implement this by submitting the entire model, validating the entire model and getting back the messages that we care about for this particular field - thought that's not shortest solution either, i feel like it might have the strongest case going forwards.